### PR TITLE
D453042: Adding file creation logic to healthcheck

### PR DESCRIPTION
**Pull request by tbroadbent22**
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=453042

### CODE REVIEW

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Has this been tested?  It doesn't look like it would work to me.  I note that I don't see "File created successfully" being logged in the build log despite the fact that I can see the Staging Service being stood up and I can see other DEBUG-level messages.

**Comment by tbroadbent22** [comment]
I wasnt sure overly sure how to test this locally given its run only once, on startup, but will look into testing it further

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        final DiskSpaceHealthIndicatorProperties diskSpaceHealthIndicatorProperties
+            = new DiskSpaceHealthIndicatorProperties();
+
+        stagingProperties = new StagingProperties();
+        stagingProperties.setHealthcheckTimeoutSeconds(60);
+
+        final File file = new File(folder.getRoot().getAbsolutePath() + "/test");
+        file.setReadOnly();
+        file.setWritable(false);
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java]: Are you sure these two file methods are having an effect. Can you add a JUnit test where this check passes.

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.springframework.boot.actuate.autoconfigure.system.DiskSpaceHealthIndicatorProperties;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.unit.DataSize;
+
+/**
+ *
+ * @author TBroadbent
+ */
+public class DiskSpaceHealthIndicatorWithTimeoutTest
+{
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java:36]: make class final

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    private StagingController controller;
+
+    BatchDao fileSystemDao;
+    HttpServletRequest request;
+    StagingProperties stagingProperties;
+    File file;
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java]: Make private, and File not needed

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    @Rule
+    public final TemporaryFolder folder = new TemporaryFolder();
+
+    @Before
+    public void setup() throws IOException
+    {
+        fileSystemDao = Mockito.mock(BatchDao.class);
+        request = Mockito.mock(HttpServletRequest.class);
+    }
+
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java:52]: No methods seem to throw IOException, if you can remove

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    @Test
+    public void healthCheckTest() throws IOException
+    {
+        final DiskSpaceHealthIndicatorProperties diskSpaceHealthIndicatorProperties
+            = new DiskSpaceHealthIndicatorProperties();
+
+        stagingProperties = new StagingProperties();
+        stagingProperties.setHealthcheckTimeoutSeconds(60);
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java:58]: You could move these two lines to setup

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java [hunk 6]
```diff
@@ -16,6 +16,14 @@
...
 
 import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
```
Comment (frankmccarry) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java]: Above 6 imports not needed.

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java (hunk 3 above)
Comment (tbroadbent22) [reply staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java]: I'm not sure these can be made final due to being initialized in setup, but can be private
image

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java (hunk 4 above)
Comment (tbroadbent22) [reply staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java:52]: Do you mean to remove the mocks?

Comment (frankmccarry) [reply staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java:52]: Your method signitures currently define  throws IOException, but they do not throw an IOException. The declared throws not required.

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java (hunk 3 above)
Comment (frankmccarry) [reply staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java]: Correct, suggested change was just make private for these.

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java [hunk 7]
```diff
@@ -54,19 +64,35 @@ protected void doHealthCheck(Health.Builder builder) throws Exception
         });
         try {
             healthcheckFuture.get(healthcheckTimeoutSeconds, TimeUnit.SECONDS);
+            testWrite(path);
```
Comment (frankmccarry) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java]: As discussed do not call directly but as a future task

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java [hunk 8]
```diff
@@ -54,19 +64,35 @@ protected void doHealthCheck(Health.Builder builder) throws Exception
...
             healthcheckFuture.cancel(true);
             builder.down().withDetail(
                 "errorMessage",
                 String.format("Exception thrown trying to access batches directory: %s", path.toString()));
             LOGGER.warn("Exception thrown trying to access batches directory {} during healthcheck", path.toString(), e);
         }
     }
+
+    private void testWrite(final File path) throws IOException
+    {
```
Comment (frankmccarry) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskSpaceHealthIndicatorWithTimeout.java]: As discussed pass in Builder so you can set up or down, also build up related error message. Also do not throw exception here.

**Review by frankmccarry — APPROVED** [review]
Changes look good, thanks.

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import java.util.concurrent.TimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.boot.actuate.health.AbstractHealthIndicator;
+import org.springframework.boot.actuate.health.Health;
+
+/**
+ *
+ * @author TBroadbent
+ */
```
Comment (dermot-hardy) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java]: Don't include `@author TBroadbent` comments in new files.  These will get stale very quickly as other people start to modify the file.  If people want to see who the author is then they can look in the source control history.

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    private final File healthcheckFile;
+    private final int healthcheckTimeoutSeconds;
+    private final ExecutorService healthcheckExecutor;
+
+    public DiskAccessHealthIndicatorWithTimeout(
+        final File path, final int healthcheckTimeoutSeconds)
+    {
+        super();
+        this.healthcheckFile = new File(path.getPath() + File.separator + "healthcheck-file.txt");
```
Comment (dermot-hardy) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java]: Don't use the legacy file I/O packages in new code.  Try to transition to the new `java.nio.file` package and then use the functionality from it.  Revert to the legacy `java.io.File` as little as possible.

So make `healthcheckFile` a `Path` rather than a `File`, and change this line to `this.healthcheckFile = path.toPath().resolve("healthcheck-file.txt");`

See [https://docs.oracle.com/javase/tutorial/essential/io/legacy.html](https://docs.oracle.com/javase/tutorial/essential/io/legacy.html).

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by frankmccarry — CHANGES_REQUESTED** [review]
Needs some rework. Filename not specified so the check is not working as you are expecting to work. Default path is "." in the code. Also the test in the check is checking for created != null, which is the opposite of what it should be testing for.

File: staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+    private final Path healthcheckFile;
+    private final int healthcheckTimeoutSeconds;
+    private final ExecutorService healthcheckExecutor;
+
+    public DiskAccessHealthIndicatorWithTimeout(
+        final Path path, final int healthcheckTimeoutSeconds)
+    {
+        super();
+        this.healthcheckFile = path.getFileName();
```
Comment (dermot-hardy) [inline staging-service/src/main/java/com/github/cafdataprocessing/services/staging/DiskAccessHealthIndicatorWithTimeout.java]: Where is the reference to `healthcheck-file.txt` now?

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ * limitations under the License.
+ */
+package com.github.cafdataprocessing.services.staging.healthcheck.test;
+
+import com.github.cafdataprocessing.services.staging.StagingController;
+import com.github.cafdataprocessing.services.staging.StagingProperties;
+import com.github.cafdataprocessing.services.staging.dao.BatchDao;
+import com.github.cafdataprocessing.services.staging.models.StatusResponse;
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
Comment (frankmccarry) [inline staging-service/src/test/java/com/github/cafdataprocessing/services/staging/healthcheck/test/DiskSpaceHealthIndicatorWithTimeoutTest.java]: unused import statement

**Review by frankmccarry — APPROVED** [review]
Changes look good. Thanks
